File: LeetCode/242-valid_anagram/242-valid_anagram.py
from typing import List


class Solution:
    # # fast
    # def isAnagram(self, s: 'str', t: 'str') -> 'bool':
    #     return ''.join(sorted(s)) == ''.join(sorted(t))

    # # slow
    # def isAnagram(self, s: 'str', t: 'str') -> 'bool':
    #     return sorted(s) == sorted(t)

    def isAnagram(self, s: 'str', t: 'str') -> 'bool':
        if len(s) != len(t):
            return False

        char_dict = {}
        for char in s:
            # if e in char_dict:
            #     char_dict[char] += 1
            # else:
            #     char_dict[char] = 1
            char_dict[char] = char_dict.get(char, 0) + 1

        for char in t:
            if char in char_dict:
                char_dict[char] -= 1
            else:
                return False

        # for k, v in char_dict.items():
        #     if v != 0:
        #         return False

        for value in char_dict.values():
            if value != 0:
                return False

        return True


if __name__ == '__main__':
    sol = Solution()
    inputs1 = ["anagram", "rat",]
    inputs2 = ["nagaram", "car",]
    outputs = [True, False]

    for idx in range(len(outputs)):
        ans = sol.isAnagram(inputs1[idx], inputs2[idx])
        if ans != outputs[idx]:
            print(f'{inputs1[idx]}, {inputs2[idx]}, {ans}, {outputs[idx]}')

File: LeetCode/242-valid_anagram/242-valid_anagram.go
package main

import (
	"fmt"
)

// // slow
// func isAnagram(s string, t string) bool {
// 	if len(s) != len(t) {
// 		return false
// 	}

// 	charMap := make(map[rune]int)

// 	for _, v := range s {
// 		charMap[v] += 1
// 	}

// 	for _, v := range t {
// 		if _, ok := charMap[v]; ok {
// 			charMap[v] -= 1
// 		} else {
// 			return false
// 		}
// 	}

// 	for _, v := range charMap {
// 		if v != 0 {
// 			return false
// 		}
// 	}

// 	return true
// }

const aPos = 97

// fast
func isAnagram(s string, t string) bool {
	if len(t) != len(s) {
		return false
	}
	list := [26]int{}

	for i := 0; i < len(s); i++ {
		list[s[i]-aPos] += 1
	}

	for i := 0; i < len(t); i++ {
		p := t[i] - aPos
		list[p] -= 1

		if list[p] < 0 {
			return false
		}
	}

	return true
}

func main() {
	inputs1 := []string{"anagram", "rat"}
	inputs2 := []string{"nagaram", "car"}
	outputs := []bool{true, false}
	var ans bool

	for pos := 0; pos < len(outputs); pos++ {

		ans = isAnagram(inputs1[pos], inputs2[pos])
		if ans != outputs[pos] {
			fmt.Println(inputs1[pos], inputs2[pos], outputs[pos], ans)
		}
	}
}
